src/api/useViewport.js

```diff
@@ -0,0 +1,85 @@
+import { useCallback } from 'react';
+import { useDispatch } from 'react-redux';
+import { setViewportFeatures } from '../redux/cartoSlice';
+import bboxPolygon from '@turf/bbox-polygon';
+import intersects from '@turf/boolean-intersects';
+
+const GEOMETRY_TYPES = Object.freeze([
+  'Point',
+  'MultiPoint',
+  'LineString',
+  'MultiLineString',
+  'Polygon',
+  'MultiPolygon'
+]);
+
+// TODO: need to find another way to get the unique feature id
+// User will need to specify the unique id as a widget prop ?¿
+// useViewport receives another argument with unique id ?¿
+function uniqueId(feature) {
+  if ('id' in feature) {
+    return feature.id;
+  }
+
+  if ('cartodb_id' in feature.properties) {
+    return feature.properties.cartodb_id;
+  }
+
+  if ('geoid' in feature.properties) {
+    return feature.properties.geoid;
+  }
+
+  throw new Error('"uniqueFeatureId" is required.');
+}
+
+function getUniqueFeatures(tiles) {
+  const features = new Map();
+
+  tiles.forEach(tile => {
+    if ('dataWithWGS84Coords' in tile) {
+      const data = tile.dataWithWGS84Coords;
+
+      for (const feature of data) {
+        const id = uniqueId(feature);
+  
+        if (!features.has(id)) {
+          features.set(id, feature);
+        }
+      }
+    }
+  });
+
+  return Array.from(features.values());
+}
+
+function featuresInViewport(features, viewport) {
+  const viewportBbox = bboxPolygon(viewport.getBounds());
+
+  return features.filter(feat => {
+    if (GEOMETRY_TYPES.includes(feat.geometry.type)) {
+      return intersects(feat, viewportBbox);
+    }
+    
+    return false;
+  });
+}
+
+export default function useViewport(source) {
```

Comment:
This hook probably be called `useViewportFeatures`, as that it what represents its content

Reply:
IMO the variable returned with the function should be call something like `calculateViewportFeatures`, not 'onViewportChange'